packages/lodestar/src/network/util.ts

```diff
@@ -52,3 +56,86 @@ export async function initializePeerInfo(peerId: PeerId, multiaddrs: string[]):
   multiaddrs.forEach((ma) => peerInfo.multiaddrs.add(ma));
   return peerInfo;
 }
+
+export function encodeChunkifyResponse(config: IBeaconConfig, method: Method, body: ResponseBody): Buffer {
+  let output= Buffer.alloc(0);
+  switch (method) {
+    case Method.Status:
+      output = serialize(config.types.Status, body);
+      return encodeSingleChunk(output);
+    case Method.Goodbye:
+      output = serialize(config.types.Goodbye, body);
+      return encodeSingleChunk(output);
+    case Method.BeaconBlocksByRange:
+    case Method.BeaconBlocksByRoot:
+      return encodeBlockChunks(config, body as BeaconBlock[]);
+  }
+}
+
+export function decodeChunkifyResponse(config: IBeaconConfig, method: Method, chunks: Buffer): ResponseBody {
+  switch (method) {
+    case Method.Status:
+    case Method.Goodbye:
+      return decodeSingleChunk(config, method, chunks) as ResponseBody;
+    case Method.BeaconBlocksByRange:
+    case Method.BeaconBlocksByRoot:
+      return decodeBlockChunks(config, method, chunks);
+  }
+}
+
+function encodeBlockChunks(config: IBeaconConfig, blocks: BeaconBlock[]): Buffer {
+  const chunkArr = blocks.map(block => encodeSingleChunk(serialize(config.types.BeaconBlock, block)));
+  return Buffer.concat(chunkArr);
```

Comment:
Maybe I'm missing something but aren't we suppose to send each chunk as soon as it's ready instead of waiting to serialize all blocks?

Reply:
yeah this is a good point. I don't see a way to do that with the current version of libp2p, looking into v0.27 of libp2p now.

Reply:
Even with new version, I think code will need reorg. You will need access to connection I guess

Reply:
yeah. I'd keep this method unchanged and reorg the code in other parts. Whenever we have a bunch of blocks, we'll call this method to get encoded Buffer and send to client.

Reply:
Not sure you can, as you are suppose to do something like:

```javascript
blocks.forEach(block => {
  connection.write(
     encodeSingleChunk(serialize(config.types.BeaconBlock, block)))
  ); 
})
```

Reply:
do you think we should write to connection/stream per block or per n blocks? I thought I'd send per n block. This depends on the way we query blocks.

Reply:
per block

Reply:
I think I found the way it works in the new libp2p. We just need to give an array of buffer to the stream and libp2p will write to the stream one by one. At the client side, it'll need to read from the stream chunk by chunk and decode each until EOF.

Reply:
If I understood you, we will again have situation where we serialize all blocks into array of buffers before sending them?

Reply:
currently I concatenate all chunks into a single Buffer. I'll change to giving an array of buffers to libp2p stream. Internally, libp2p will send each buffer/chunk one by one.  I checked at the client side, it received chunks one by one like:
```
received from server: one
received from server: two
```

Reply:
But we will still wait for all blocks to serialize before sending them?

Reply:
depends on the how our db api returns blocks.
If we could yield block by block, we could serialize individually, then push each block one by one into the stream.
If the db just returns an array of blocks, then it probably just makes sense to map(serialize) and push the entire array at once

Reply:
if we don't want to serialize all blocks before sending to stream, we can use something like https://github.com/alanshaw/it-pipe , that's the lib I see in every example of the new libp2p.
Server will get one block => serialize => send to stream => client gets that serialized data => deserialize => server get next block => serialize => send to stream => ...
I think this clears everybody's concern.

Reply:
take a look at https://github.com/alanshaw/it-awesome too, plenty of tools to pick from

Reply:
We could expose stream method since database is returning blocks one-by-one as default behavior (we are just waiting for all of them. But I think priority is just sending blocks while serializing as I'm pretty sure it has more significant performance impact

Reply:
I got your point